tket2/src/serialize.rs

```diff
@@ -1,13 +1,421 @@
+    ///
+    /// If the circuit is not a function in a module-rooted HUGR, a new module
+    /// is created with a `main` function containing it.
+    ///
+    /// # Errors
+    ///
+    /// - If the circuit's parent is not the root of the HUGR or a function in
+    ///   the root's module.
+    ///
+    // TODO: Store the path pointer on `METADATA_ENTRYPOINT` instead.
+    pub fn to_package_writer(&self, writer: impl io::Write) -> Result<(), CircuitStoreError> {
+        let hugr = self.hugr().as_ref();
+
+        // Check if we support storing the circuit as a package.
+        //
+        // This restriction may be relaxed once `METADATA_ENTRYPOINT` is implemented.
+        let circuit_is_at_root = self.parent() == hugr.root();
+        let circuit_is_fn_at_module_root = OpTag::ModuleRoot
+            .is_superset(hugr.get_optype(hugr.root()).tag())
+            && hugr.get_parent(self.parent()) == Some(hugr.root());
+        if !circuit_is_at_root && !circuit_is_fn_at_module_root {
+            return Err(CircuitStoreError::NonRootCircuit {
+                parent: self.parent(),
+            });
+        }
+
+        let mut pkg = Package::from_hugr(hugr.clone())?;
+        pkg.extensions = self.required_extensions().to_owned();
+
+        Ok(pkg.to_json_writer(writer)?)
+    }
+}
+
+impl Circuit<Hugr> {
+    /// Load a circuit from a package or hugr json.
+    ///
+    /// If the json encodes a package, one of the modules must contain a
+    /// function called `function_name`.
+    ///
+    /// Otherwise, the json must encode a module-rooted HUGR containing the
+    /// named function.
+    pub fn load_function_reader(
+        json: impl io::Read,
+        function_name: impl AsRef<str>,
+    ) -> Result<Self, CircuitLoadError> {
+        let mut pkg = Package::from_json_reader(json)?;
+        pkg.update_validate(&mut REGISTRY.clone())?;
+        let Package {
+            modules,
+            extensions,
+        } = pkg;
+
+        let (_module_idx, mut circ) = find_function_in_modules(modules, function_name.as_ref())?;
+        if !extensions.is_empty() {
+            circ.set_required_extensions(Arc::new(extensions));
+        }
+        Ok(circ)
+    }
+
+    /// Load a circuit from a hugr json.
+    ///
+    /// The circuit points to the Hugr's root or, if the `TKET2.entrypoint` metadata is present,
+    /// the indicated node.
+    ///
+    /// # Errors
+    ///
+    /// - If the target circuit root is not a dataflow container.
+    pub fn load_hugr_reader(json: impl io::Read) -> Result<Self, CircuitLoadError> {
+        let mut hugr: Hugr = serde_json::from_reader(json)?;
+        hugr.update_validate(&REGISTRY.clone())?;
+        // TODO: Read the entrypoint from the metadata.
+        let root = hugr.root();
+        Ok(Circuit::try_new(hugr, root)?)
+    }
+}
+
+/// Error type for conversion between `Op` and `OpType`.
+#[derive(Debug, Display, Error, From)]
+#[non_exhaustive]
+pub enum CircuitStoreError {
```

Comment:
when are we dropping the word "Circuit" 👀

Reply:
Soon™️
(it should be my next PR)